README.md
# Youtube2MP3
Beta version of youtube-mp3 GUI program

Requirements:

~~Python installed and added to the path~~
#
# -New release uses libcUrl to download the files
# -See latest release if you do not wish to install python
# 
youtube-dl added to the path
ffmpeg added to the path

external headers:
~~picosha2 for SHA256 IDs~~ #Now redundant after libcUrl patch
json for modern C++ (json.hpp) for parsing the youtube JSON API

# Changes for 0.3:

1. Added generalised error boxes to help show where errors may lie
2. Added functionality for importing entire playlists through youtube
3. Added more poorly written comments
4. Fixed a major bug causing QString to const char* conversions to fail...
 therefore causing the application to incorrectly download youtube json and or incorrectly store it in the wrong file
5. Fixed a bug causing the DataWidgets to be initially shown in the wrong order
6. Added font scaling so titles and urls will now correctly fit on the screen

# Hopeful upcoming fixes

1. Add a progress box for downloading files
 (remnants of this are already in the source code. I am currently unable to get a responsive UI while downloading the songs)
2. Additional options for displaying DataWidgets moire compactly

currently very early beta: more updates soon

Will be lots of errors please inform me of them so I can fix them ty <3


changes.txt
Changes:

Added done dialog after files downloaded
Song downloads from a playlist download into one file under ther playlist id name

main.cpp
#include "basewindow.h"
#include "filedownload.h"
#include "json.hpp"
#include <QtCore>
#include <QtWidgets>
#include <QApplication>
#include <Windows.h>

QString getYoutubeID(QString Url)
{
    QRegularExpressionMatch IDMatch = idReg.match(Url);
    return IDMatch.captured(0);
}

QString getPlaylistID(QString Url) 
{
    QRegularExpressionMatch IDMatch = playlistIDReg.match(Url);
    return IDMatch.captured(0);
}

QString getLength(std::string ISOString)
{
    //Get the QString in format "(m):(s)" from ISO 8601 duration string. Does not accept hours
    QString ISOQString = QString::fromStdString(ISOString);
    ISOQString.remove("PT").remove("S");
    QStringList Times = ISOQString.split("M");
    QString Minutes = Times.value(0);
    QString Seconds = Times.value(1);
    if (Seconds.toInt() < 10){
        Seconds = "0"+Seconds;
    }
    QString formatTime = Minutes + ":" + Seconds;
    return formatTime;
}

void execCmd(const char* command)
{
    //Command to open up cmd without causing a window pop up
    //Used to run youtube-dl and pass arguments
    wchar_t *path = (wchar_t*) malloc(sizeof(wchar_t) * 500);
    for (int i = 0; i < strlen(command) + 1; i++)
    {
        path[i] = command[i];
    }

    PROCESS_INFORMATION pi;
    STARTUPINFO si;

    ZeroMemory(&si, sizeof(si));
    si.cb = sizeof(si);
    ZeroMemory(&pi, sizeof(pi));

    if (!CreateProcess(NULL,
        path,
        NULL,
        NULL,
        FALSE,
        CREATE_NO_WINDOW,
        NULL,
        NULL,
        &si,
        &pi))
    {
        qDebug() << "Could not start command "<< QString::fromLocal8Bit(command) << GetLastError();
        return;
    }
    WaitForSingleObject(pi.hProcess, INFINITE);

    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
}

bool dirExists(const std::string& dirName_in)
{
  //Check if the directory has been made by the JSON download
  //Thus checks vailidity of URL
  DWORD ftyp = GetFileAttributesA(dirName_in.c_str());
  if (ftyp == INVALID_FILE_ATTRIBUTES)
    return false;  //Path error

  if (ftyp & FILE_ATTRIBUTE_DIRECTORY)
    return true;   //Directory exists

  return false;    //Directory not found
}

void createErrorMessage(QString Error) 
{
    QMessageBox* errorBox = new QMessageBox();
    errorBox->setIcon(QMessageBox::Warning);
    errorBox->setWindowTitle("ERROR");
    errorBox->setText("The program has encountered an error:");
    errorBox->setInformativeText(Error);
    errorBox->setStandardButtons(QMessageBox::Abort);
    QObject::connect(errorBox, &QMessageBox::buttonClicked, errorBox, &QMessageBox::close);
    errorBox->exec();
}

QFont scaleFont(QFont Font, QString text, int width, int height)
{
    //Iterate through to find appropriate font to represent the text in the label
    QLabel Label;
    Label.setFont(Font);
    Label.resize(width, height);
    QRect bound = QFontMetrics(Font).boundingRect(
                0, 0, 320, 180, Qt::AlignCenter, text);
    bool fit = (bound.width() <= Label.width() &&
                bound.height() <= Label.height());
    while (!fit) {
        Font.setPointSize(Font.pointSize() - 1);
        QRect bound = QFontMetrics(Font).boundingRect(
                    0, 0, 320, 180, Qt::AlignCenter, text);
        if ((bound.width() <= Label.width() &&
             bound.height() <= Label.height())) {
            return Font;
        }
    }
    return Font;

}

void SetSize(QWidget *window, int x,int y)
{
    //Force QWidget size
    window->resize(x,y);
    window->setMaximumSize(x,y);
    window->setMinimumSize(x,y);
}

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);
    QApplication::setApplicationName("Youtube2MP3");
    qDebug() << "Version 0.3.1";

    BaseWindow MainWindow;
    MainWindow.show();
    MainWindow.setAttribute(Qt::WA_QuitOnClose);

    return app.exec();
}




datalabel.h
#ifndef DATALABEL_H
#define DATALABEL_H

#include <QtCore>
#include <QtGui>
#include <QtWidgets>

class DataLabel
{
public:
    DataLabel(QPixmap Small, QPixmap Large, QString Name,
              QString url, QString id, QString directory,
              QString duration, QString outdir, int Length); //Set up of class
    QPixmap LogoSmall; //Small thumbnail
    QPixmap LogoLarge; //LargeThumbnail
    QString Title; //Video title
    QString Url; //Link to video
    QString ID; //Video ID
    QString Directory; //Location of folder in which JSON and thumbails are stored
    QString Duration; //Formatted time (m):(s)
    QString Outdir;
    int length; //Time in seconds
};

class DataWidget : public QWidget
{
    // Widget to show the current inputted videos awaiting download
    Q_OBJECT

public:
    DataWidget(QWidget *parent, DataLabel* data);
    DataLabel* VidData;

public slots:
    void NextPressed();
    void DeletePressed();

private:
    QLabel* PictureLabel;
    QLabel* titleLabel;
    QLabel* timeLabel;
    QLabel* urlLabel;
    QPushButton* DelButton;
    QPushButton* NextButton;

signals:
    void DeleteThis(QString ID);
    void NextData(QString ID);
};

#endif // DATALABEL_H


basewindow.h
#ifndef BASEWINDOW_H
#define BASEWINDOW_H

#include "datalabel.h"
#include "json.hpp"
#include <QtCore>
#include <QtWidgets>
#include <QApplication>

//Constants
const QRegularExpression idReg("((?<=(v|V)/)|(?<=be/)|(?<=(\\?|\\&)v=)|(?<=embed/))([\\w-]++)"); //Regex to extract youtube id from any url
const QRegularExpression playlistIDReg("((?<=(<)/)|(?<=(\\?|\\&)list=)|(?<=</))([\\w-]++)"); //Regex to extract playlist id from url
//Flags to pass to GetUrl function
const int SUBMIT_VIDEO = 1; 
const int SUBMIT_PLAYLIST = 2;

class Downloads : public QObject
{
    //Worker class to download the files.
    Q_OBJECT

public:
    Downloads(QVector<DataWidget*> Files, std::string Directory);
    void DownloadFiles();

private:
    QVector<DataWidget*> DLFiles;
    std::string directory;

signals:
    void finished();
    //void finSong(QString title);
};

class BaseWindow : public QWidget
{

    Q_OBJECT

public:
    BaseWindow(QWidget *parent = 0);
    QString Directory = qApp->applicationDirPath()+"/"; //Main app directory
    nlohmann::json ParseJson(std::fstream& DataFile); //Returns the parsed json from the given file
    void SubmitUrl(QString ytID, int outFlag, QString outdir); //Add file to download list and create widget
    QMenuBar *WindowMenu;
    QMenu *MainMenu;
    QAction *InputUrl;
    QAction *PlaylistInputUrl;
    QAction *DownloadAction;
    QAction *QuitAction;
    ~BaseWindow();

private:
    void CreateActions(); //Create all MainWindow menu actions
    void LoadWidgets(); //Load all created widgets for MainWindow
    void createNewData(QString ID, int outFlag, QString PlaylistID); //Create new data label and data widget for submitted video
    void GetUrl(int Flag); //Open QDialog for inputting an url and begin procedure of adding url to list
    void showLastWidget();
    QString CreateJSONUrl(QString ytID); //Create video JSON url using youtube api v3
    QString CreatePlaylistUrl(QString plID); //Create playlist JSON url using youtube api v3
    QDialog *UrlBox;
    QPushButton *SubmitButton;
//    QDialog *ProgressBox = new QDialog();
//    QLabel *ProgLabel = new QLabel("", ProgressBox);
    QLineEdit *UrlInput;
    QVector<DataWidget*> DataWidgets; //Vector containing the data widgets
    const QString apiUrl = "https://www.googleapis.com/youtube/v3/videos?part=snippet,contentDetails&id=%1&key=%2";
    const QString apiPLUrl = "https://www.googleapis.com/youtube/v3/playlistItems?part=contentDetails&maxResults=50&playlistId=%1&key=%2";
    const QString apiKey = "";
    const QString baseOutDir = Directory+"Bin/Data/";

public slots:
    void CallGetUrlForVideo();
    void CallGetUrlForPlaylist();
    void GetIDFromUrlBox();
    void SubmitPlaylistUrl();
    void DeleteData(QString MatchID); //Deletes the DataWidget and DataLabel of the given ID
    void ShowNext(QString ID); //Shows the next DataWidget in list
    void DownloadMP3(); //Downloads the MP3 file using cmd line youtube-dl
    void CleanUpAll(); //Delete all DataWidgets
//  void ShowDLProgress(QString title); //Updates the ProgressBox on download
};


//Utility functions (definitely not taken from stack overflow at all)
QString getLength(std::string ISOString);
void execCmd(const char* command);
bool dirExists(const std::string& dirName_in);
void SetSize(QWidget* window, int x,int y);
void createErrorMessage(QString Error);
QString getYoutubeID(QString Url); //Use the constant regex to extract yt video ID
QString getPlaylistID(QString Url); //Use the constant regex to extract the playlist ID
QFont scaleFont(QFont Font, QString text, int width, int height); //Scale font to properly fit a labels size

#endif // BASEWINDOW_H


datalabel.cpp
#include "datalabel.h"
#include "basewindow.h"
#include <QtCore>
#include <QtWidgets>
#include <QApplication>

DataLabel::DataLabel(QPixmap Small, QPixmap Large, QString Name,
                     QString url, QString id, QString directory,
                     QString duration, QString outdir, int Length)
{
    LogoSmall = Small;
    LogoLarge = Large;
    Title = Name;
    Url = url;
    ID = id;
    Directory = directory;
    Duration = duration;
    Outdir = outdir;
    length = Length;
}

DataWidget::DataWidget(QWidget *parent, DataLabel* data)
    : QWidget(parent)
{
    //Initialising the widget and all components
    VidData = data;
    setStyleSheet("QWidget {background-color: rgb(230, 230, 240);}");
    move(0, 24);
    setFont(QFont("Arial",25));
    SetSize(this, 730, 408);
    PictureLabel = new QLabel(this);
    PictureLabel->setPixmap(VidData->LogoLarge);
    PictureLabel->setMask(VidData->LogoLarge.mask());
    PictureLabel->setGeometry(205,0, 320,180);
    PictureLabel->show();
    titleLabel = new QLabel(VidData->Title, this);
    titleLabel->setGeometry(0,190, 730,45);
    titleLabel->setFont(scaleFont(QFont("Arial", 20), VidData->Title, 730, 45));
    titleLabel->setAlignment(Qt::AlignCenter);
    titleLabel->show();
    timeLabel = new QLabel(VidData->Duration,this);
    timeLabel->setFont(QFont("Arial",20));
    timeLabel->setGeometry(0,235, 730,40);
    timeLabel->setAlignment(Qt::AlignCenter);
    timeLabel->show();
    urlLabel = new QLabel("Url: "+VidData->Url, this);
    urlLabel->setFont(scaleFont(QFont("Arial", 20), VidData->Url, 730, 45));
    urlLabel->setGeometry(0,280, 730,45);
    urlLabel->setAlignment(Qt::AlignCenter);
    urlLabel->show();
    DelButton = new QPushButton("Delete", this);
    DelButton->setFont(QFont("Arial",20));
    DelButton->setStyleSheet("QPushButton {"
                             "background-color: rgb(130, 130, 130);"
                             "color: rgb(200, 200, 200);}"
                             "QPushButton::pressed {"
                             "background-color: rgb(110, 110, 110);}");
    DelButton->setGeometry(110,330, 200,60);
    DelButton->show();
    NextButton = new QPushButton("Next", this);
    NextButton->setFont(QFont("Arial", 20));
    NextButton->setStyleSheet("QPushButton {"
                             "background-color: rgb(130, 130, 130);"
                             "color: rgb(200, 200, 200);}"
                             "QPushButton::pressed {"
                             "background-color: rgb(110, 110, 110);}");
    NextButton->setGeometry(420,330, 200,60);
    NextButton->show();
    connect(DelButton, &QPushButton::released, this, &DataWidget::DeletePressed);
    connect(NextButton, &QPushButton::released, this, &DataWidget::NextPressed);
}

void DataWidget::NextPressed()
{
    emit NextData(VidData->ID);
}

void DataWidget::DeletePressed()
{
    emit DeleteThis(VidData->ID);
}


filedownload.cpp
/* File source code pretty much copied from:
 * https://curl.haxx.se/libcurl/c/url2file.html
*/

#include <curl/curl.h>
#include <stdio.h>
#include <stdlib.h>
#include <QString>
#include <QDebug>

size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream) {
    size_t written = fwrite(ptr, size, nmemb, stream);
    return written;
}

int GetFile(QString outfile, QString url)
{ // Simple libcUrl function to download files from the web
    CURL *curl;
    FILE *fp;
    CURLcode res;
    curl = curl_easy_init();
    if (curl) {
        QByteArray outfileBArray = outfile.toUtf8();
        const char* outfileCStr = outfileBArray.constData();
        QByteArray urlBArray = url.toUtf8();
        const char* urlCStr = urlBArray.constData();
        errno_t err;
        err = fopen_s(&fp, outfileCStr, "wb");
        if (err == 0) {
            qDebug() << "opened";
        } else {
            qDebug() << "error";
        }
        curl_easy_setopt(curl, CURLOPT_URL, urlCStr);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0);
        res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        if(fp != NULL) {
            fclose(fp);
        }
    }
    else {
        return 0;
    }
    return 1;
}


basewindow.cpp
#include "basewindow.h"
#include "datalabel.h"
#include "filedownload.h"
#include "json.hpp"
#include <QtCore>
#include <QtWidgets>
#include <QApplication>
#include <cstdlib>
#include <fstream>
#include <string>
#include <Windows.h>


using json = nlohmann::json;

Downloads::Downloads(QVector<DataWidget*> Files, std::string Directory)
{
    DLFiles = Files;
    directory = Directory;
}

void Downloads::DownloadFiles()
{
    //Loops through each element in the vector executing the download command
    for (int i=0; i < DLFiles.length(); i++) {
        DataLabel* FileData = DLFiles[i]->VidData;
        QString escapedTitle = FileData->Title.replace("/", "_").replace("\\","-");
        QString escapedOutdir = FileData->Outdir.replace("/", "_").replace("\\","-");
        std::string OUTFILE = "-o \""+directory+"Downloads/"+escapedOutdir.toStdString()+"/"
                +escapedTitle.toStdString()+".%(ext)s\"";
        qDebug() << QString::fromStdString(OUTFILE);
        std::string DLCommand = "youtube-dl --extract-audio --audio-format mp3 --audio-quality 0 "
                +OUTFILE+" "+FileData->Url.toStdString();
        qDebug() << QString::fromStdString(DLCommand);
        execCmd(DLCommand.c_str());
        //emit finSong(FileData->Title);
    }
    QMessageBox doneDialog;
    doneDialog.setWindowTitle("Download finished");
    doneDialog.setIcon(QMessageBox::Information);
    doneDialog.setText("Downloads finished successfully");
    doneDialog.setInformativeText("Downloaded files");
    doneDialog.setStandardButtons(QMessageBox::Ok);
    connect(&doneDialog, &QMessageBox::buttonClicked, &doneDialog, &QMessageBox::close);
    doneDialog.exec();
    emit finished();
}

BaseWindow::BaseWindow(QWidget *parent)
    : QWidget(parent)
{
    //Main window set up
    if(!QDir(Directory+"Bin"+"/Data").exists()) {
        QDir().mkdir(Directory+"Bin"+"/Data");
    }
    setWindowTitle(QString("Youtube2MP3"));
    SetSize(this, 730, 420);
    setStyleSheet("QWidget {background-color: rgb(230, 230, 240);}");
    CreateActions();
    LoadWidgets();
}

void BaseWindow::CleanUpAll()
{
    //Function to remove all elements from the vector and hide them
    while (!(DataWidgets.isEmpty())) {
        DataWidgets[0]->hide();
        DataWidgets.remove(0);
    }
}

//void BaseWindow::ShowDLProgress(QString title)
//{
//    ProgLabel->setText("Finished downloading: " + title);
//}

void BaseWindow::DownloadMP3()
{
    //Start the download process for all files
    Downloads* AllFiles = new Downloads(DataWidgets, Directory.toStdString());
    QThread* DLThread = new QThread();
    AllFiles->moveToThread(DLThread);
    connect(AllFiles, &Downloads::finished, this, &BaseWindow::CleanUpAll);
//    connect(AllFiles, &Downloads::finSong, this, &BaseWindow::ShowDLProgress);

//    //Set up progress dialog
//    ProgressBox->setWindowFlags(Qt::WindowSystemMenuHint | Qt::WindowTitleHint | Qt::WindowCloseButtonHint);
//    SetSize(ProgressBox, 550, 150);
//    ProgressBox->setWindowTitle("Current Progress");
//    ProgressBox->setModal(true);
//    ProgressBox->setFont(QFont("Arial", 18));
//    ProgressBox->setStyleSheet("QDialog {background-color: rgb(230, 230, 240);}");
//    ProgLabel->setGeometry(0, 20, 550, 60);
//    ProgLabel->setAlignment(Qt::AlignCenter);
//    ProgLabel->show();

//    //progress dialog set up end

    DLThread->start();
    AllFiles->DownloadFiles();
    //ProgressBox->close();
}

void BaseWindow::DeleteData(QString MatchID)
{
    //Loops through to find a matching ID then hiding it and deleting it from the Vector
    for (int i = 0; i < DataWidgets.count(); i++) { //Loop through data widgets
        if (DataWidgets[i]->VidData->ID == MatchID) { //When found the matching id
            DataWidgets[i]->hide();
            DataWidgets.remove(i);
            if (DataWidgets.count() == 0){ //Return if no data left
                break;
            }
            else {
                if (i == DataWidgets.count()) { //If deleted widget was last in list
                    DataWidgets[i-1]->show(); //Show then new last element
            } else {
                    DataWidgets[i]->show();
                }
            }
        }
        else {

        }
    }
}

void BaseWindow::ShowNext(QString ID)
{
    if ( DataWidgets.count() <= 1) {return;}
    //Loops through for a matching ID then shows the widget
    qDebug() << "Data widgets count = " << DataWidgets.count();
    for (int i = 0; i < DataWidgets.count(); i++) {
        if (DataWidgets[i]->VidData->ID == ID){ //When found matching id
            qDebug() << i;
            DataWidgets[i]->hide();
            if (i == DataWidgets.count() -1){
                qDebug() << "load start";
                DataWidgets[0]->show(); //Load initial data widget
                break;
            }
            else {
                qDebug() << "load next one";
                DataWidgets[i+1]->show();
                break;
            }
        }
        else {

        }
    }
}

void BaseWindow::showLastWidget()
{
    //Used to ensure only one Data Widget is currently open.
    if (DataWidgets.count() == 0) {
        return;
    }
    for (int i = 0; i < DataWidgets.count(); i++) {
        DataWidgets[i]->hide();
    }
    DataWidgets[DataWidgets.count()-1]->show();
}

json BaseWindow::ParseJson(std::fstream& DataFile)
{
    if (DataFile.fail()){
        qDebug() << "Failed to load" << strerror(errno);
        createErrorMessage("Failed to load " + QString::fromLocal8Bit(strerror(errno)));
        return NULL;
    }
    qDebug() << "File Loaded";
    json Data;
    //Check if JSON can be parsed into data
    try{
        Data = json::parse(DataFile);
    } catch (json::parse_error &e){
        createErrorMessage("JSON File cannot be parsed - " + QString::fromLocal8Bit(e.what()));
        qDebug() << QString::fromLocal8Bit(e.what());
        return NULL;
    }
    return Data;
}

void BaseWindow::createNewData(QString ID, int outFlag, QString PlaylistID = NULL)
{
    //Create the data for a new datalabel
    if (!dirExists(Directory.toStdString()+"/Bin/Data/"+ID.toStdString())) { //check download
        createErrorMessage("Could not find the downloaded song metadata.");
        return;
    }
    
    QString Folder = QString(Directory+"/Bin/Data/"+ID); //Save folder directory
    std::fstream DataFile(QString(Directory+"/Bin/Data/"+ID+"/Info.json").toStdString());
    json Data = ParseJson(DataFile);
    if (Data == NULL) {
        return;
    }
    QString title = QString::fromStdString(Data["items"][0]["snippet"]["title"]);
    std::string ISOLength = Data["items"][0]["contentDetails"]["duration"];
    QString url = "https://youtu.be/"+QString::fromStdString(Data["items"][0]["id"]); //Remove extra tags from url to stop youtube-dl doing extra stuff
    if (QString::fromStdString(ISOLength).contains("H")){
        createErrorMessage("Video too long. Videos must be less than one hour.");
        return;
    }
    QString FormatLength = getLength(ISOLength); //Get formattted length from JSON
    QStringList MinsSecs = FormatLength.split(":");
    QString minutes = MinsSecs.value(0);
    QString secs = MinsSecs.value(1);
    int secsDuration = minutes.toInt() * 60 + secs.toInt(); //Get length of song from formatted length
    std::string smallPic = Data["items"][0]["snippet"]["thumbnails"]["default"]["url"]; //Urls for thumbnails to be downloaded from
    std::string bigPic = Data["items"][0]["snippet"]["thumbnails"]["medium"]["url"];
    if (GetFile((baseOutDir+ID+"/smallThumbnail.jpg").toStdString().c_str(), smallPic.c_str()) == 1 
    && GetFile((baseOutDir+ID+"/bigThumbnail.jpg").toStdString().c_str(), bigPic.c_str()) == 1) {
     //Download both small and big thumbnails
    } else {
        createErrorMessage("Error downloading thumbnails.");
        return;
    }
    QString outdir;
    if (outFlag == 1 && !PlaylistID.isNull()) {
        outdir = PlaylistID;
    } else {
        outdir = title;
    }
    if (!QDir(Directory + "Downloads/" + outdir).exists()) {
        QDir().mkdir(Directory + "Downloads/" + outdir);
    }
    QPixmap logoPix(baseOutDir+ID+"/smallThumbnail.jpg");
    QPixmap highPix(baseOutDir+ID+"/bigThumbnail.jpg");
    DataLabel* NewData = new DataLabel(logoPix, highPix, title, url, ID,
                                       Folder, FormatLength, outdir, secsDuration); //Create the new data
    DataWidget* NewWidget = new DataWidget(this, NewData);
    DataWidgets.append(NewWidget); //Add the new data to the storage
    connect(NewWidget, &DataWidget::DeleteThis, this, &BaseWindow::DeleteData);
    connect(NewWidget, &DataWidget::NextData, this, &BaseWindow::ShowNext);
    showLastWidget();
}

void BaseWindow::SubmitPlaylistUrl() 
{
    QString Url = UrlInput->text();
    UrlBox->close();
    if (Url.isEmpty()) {
        createErrorMessage("No url given.");
        return;
    }
    QString playlistID = getPlaylistID(Url);
    if (playlistID.isEmpty()) {
        createErrorMessage("Could not determine playlist ID from url");
    }
    QString outFile = (baseOutDir + playlistID + "/playlistInfo.json");
    std::ofstream createFile(outFile.toStdString());
    createFile.close();
    if (!QDir(baseOutDir + playlistID).exists()) {
        QDir().mkdir(baseOutDir + playlistID);
    } 
    
    if(GetFile(outFile, CreatePlaylistUrl(playlistID)) == 1){

    } else {
        createErrorMessage("Error downloading playlist metadata.");
        return;
    }

//    if (!dirExists(outFile.toStdString())) {
//        createErrorMessage("Could not find downloaded metadata.");
//        return;
//    }
    std::fstream DataFile(outFile.toStdString());
    json Data = ParseJson(DataFile);
    if (Data == NULL) {
        return;
    }
//    QDialog ProgressBox;
//    ProgressBox.setWindowFlags(Qt::WindowSystemMenuHint | Qt::WindowTitleHint | Qt::WindowCloseButtonHint);
//    SetSize(&ProgressBox, 550, 150);
//    ProgressBox.setWindowTitle("Current Progress");
//    ProgressBox.setModal(true);
//    ProgressBox.setFont(QFont("Arial", 18));
//    ProgressBox.setStyleSheet("QDialog {background-color: rgb(230, 230, 240);}");
//    QLabel *DlProgLabel = new QLabel("", &ProgressBox);
//    DlProgLabel->setGeometry(0, 20, 550, 60);
//    DlProgLabel->setAlignment(Qt::AlignCenter);
//    ProgressBox.show();
    //For loop to pass all given IDs in the playlist to the submiturl function
    for (int i=0; i<Data["pageInfo"]["totalResults"]; i++) {
        QString ID = QString::fromStdString(Data["items"][i]["contentDetails"]["videoId"]);
        //ProgLabel->setText("Now parsing video with ID:" + ID);
        //ProgLabel->show();
        SubmitUrl(ID, 1, playlistID);
        //ProgLabel->hide();
    }

}

void BaseWindow::SubmitUrl(QString ytID, int outFlag, QString outdir)
{
    //Download dedicated JSON and create new data widget to be downloaded.
    QString outFile = (baseOutDir + ytID + "/Info.json");
    std::ofstream createFile(outFile.toUtf8().constData());
    createFile.close();
    QString URL = CreateJSONUrl(ytID);
    //qDebug() << outFile;
    if (!QDir(baseOutDir+"/"+ytID).exists()) {
        qDebug() << "making directory";
        QDir().mkdir(baseOutDir+"/"+ytID);
    }
    //qDebug() << outFile;
    if (GetFile(outFile, URL) == 1) {
    } else {
        createErrorMessage("Error downloading video metadata.");
        return;
    }
    //qDebug() << URL;
    createNewData(ytID, outFlag, outdir); //Turn JSON into datalabel
}

QString BaseWindow::CreatePlaylistUrl(QString plID)
{
    return apiPLUrl.arg(plID, apiKey).toLocal8Bit();
}
QString BaseWindow::CreateJSONUrl(QString ytID)
{
    return apiUrl.arg(ytID, apiKey);
}

void BaseWindow::GetIDFromUrlBox()
{
    //Retrieve the UrlBox submitted text and pass it to submiturl function
    QString Url = UrlInput->text();
    UrlBox->close();
    if (Url.isEmpty()){
        createErrorMessage("No url given.");
        return;
    }
    QString ID = getYoutubeID(Url);
    if (ID.isEmpty()) {
        createErrorMessage("Could not determine video ID");
        return;
    }
    SubmitUrl(ID, 0, NULL);
}

void BaseWindow::GetUrl(int Flag)
{
    //Create window for user to submit an url
    UrlBox = new QDialog();
    UrlBox->setWindowFlags(Qt::WindowSystemMenuHint | Qt::WindowTitleHint | Qt::WindowCloseButtonHint);
    SetSize(UrlBox, 550, 200);
    UrlBox->setWindowTitle("Input url");
    UrlBox->setModal(true);
    UrlBox->setFont(QFont("Arial", 18));
    UrlBox->setStyleSheet("QDialog {background-color: rgb(230, 230, 240);}");
    QLabel *UrlLabel = new QLabel("Url: ", UrlBox);
    UrlLabel->setGeometry(0,10, 550, 40);
    UrlLabel->setAlignment(Qt::AlignCenter);
    UrlLabel->show();
    UrlInput = new QLineEdit(UrlBox);
    UrlInput->setGeometry(80,60, 390,35);
    UrlInput->setAlignment(Qt::AlignCenter);
    UrlInput->show();
    SubmitButton = new QPushButton("Submit", UrlBox);
    SubmitButton->setGeometry(220,110, 110,50);
    SubmitButton->setStyleSheet("QPushButton {"
                                "background-color: rgb(130, 130, 130);"
                                "color: rgb(200, 200, 200);}"
                                "QPushButton::pressed {"
                                "background-color: rgb(110, 110, 110);}");
    SubmitButton->show();
    if (Flag == 1) {
        connect(SubmitButton, &QPushButton::released, this, &BaseWindow::GetIDFromUrlBox);
    } else if (Flag == 2) {
        connect(SubmitButton, &QPushButton::released, this, &BaseWindow::SubmitPlaylistUrl);
    } else {
        connect(SubmitButton, &QPushButton::released, UrlBox, &QDialog::close);
    } //Should never run
    UrlBox->exec();
}

void BaseWindow::CallGetUrlForVideo() {GetUrl(SUBMIT_VIDEO);}

void BaseWindow::CallGetUrlForPlaylist() {GetUrl(SUBMIT_PLAYLIST);}

void BaseWindow::CreateActions()
{
    //Create actions for the menu bars
    InputUrl = new QAction(tr("&Import url"), this);
    InputUrl->setShortcut(QKeySequence::New);
    InputUrl->setToolTip("Import video from youtube url");
    PlaylistInputUrl = new QAction(tr("&Import playlist"), this);
    PlaylistInputUrl->setShortcut(Qt::CTRL + Qt::SHIFT + Qt::Key_N);
    PlaylistInputUrl->setToolTip("Import videos from youtube playlist");
    DownloadAction = new QAction(tr("&Download files"), this);
    DownloadAction->setShortcut(QKeySequence::Save);
    DownloadAction->setToolTip("Download the inputted files");
    QuitAction = new QAction(tr("&Quit"), this);
    QuitAction->setShortcut(QKeySequence::Quit);
    QuitAction->setToolTip("Close the application");
    connect(InputUrl, &QAction::triggered, this, &BaseWindow::CallGetUrlForVideo);
    connect(PlaylistInputUrl, &QAction::triggered, this, &BaseWindow::CallGetUrlForPlaylist);
    connect(DownloadAction, &QAction::triggered, this, &BaseWindow::DownloadMP3);
    connect(QuitAction, &QAction::triggered, this, &QApplication::quit);
    if (!QDir(baseOutDir).exists()) {
        QDir().mkdir(baseOutDir);
    }
}

void BaseWindow::LoadWidgets()
{
    //Load main widgets with stylesheets
    WindowMenu = new QMenuBar(this);
    MainMenu = new QMenu("&File");
    MainMenu->addAction(InputUrl);
    MainMenu->addSeparator();
    MainMenu->addAction(PlaylistInputUrl);
    MainMenu->addSeparator();
    MainMenu->addAction(DownloadAction);
    MainMenu->addSeparator();
    MainMenu->addAction(QuitAction);
    WindowMenu->addMenu(MainMenu);
    WindowMenu->setFont(QFont("Arial", 12));
    WindowMenu->setStyleSheet("QMenuBar {"
                              "background-color: rgb(130, 130, 130);"
                              "color: rgb(200,200,200);}"
                              "QMenuBar::item {"
                              "background-color: rgb(130, 130, 130);"
                              "color: rgb(200, 200, 200);}"
                              "QMenuBar::item::selected {"
                              "background-color: rgb(90, 90, 90);}");
    MainMenu->setStyleSheet("QMenu {"
                            "background-color:rgb(130, 130, 130);"
                            "color: rgb(200,200,200);"
                            "border: 1px solid;}"
                            "QMenu::item::selected {"
                            "background-color: rgb(90, 90,90);}");
    WindowMenu->show();
}

BaseWindow::~BaseWindow()
{

}
